buffer/src/main/java/io/netty/buffer/api/BufferAllocator.java

```diff
@@ -0,0 +1,167 @@
+     * <p>
+     * The buffer will use the current platform native byte order by default, for accessor methods that don't have an
+     * explicit byte order.
+     *
+     * @param size The size of {@link Buffer} to allocate.
+     * @return The newly allocated {@link Buffer}.
+     */
+    Buffer allocate(int size);
+
+    /**
+     * Allocate a {@link Buffer} of the given size in bytes. This method may throw an {@link OutOfMemoryError} if there
+     * is not enough free memory available to allocate a {@link Buffer} of the requested size.
+     * <p>
+     * The buffer will use the given byte order by default.
+     *
+     * @param size The size of {@link Buffer} to allocate.
+     * @param order The default byte order used by the accessor methods that don't have an explicit byte order.
+     * @return The newly allocated {@link Buffer}.
+     */
+    default Buffer allocate(int size, ByteOrder order) {
+        return allocate(size).order(order);
+    }
+
+    /**
+     * Create a supplier of "constant" {@linkplain Buffer Buffers} from this allocator, that all have the given
+     * byte contents. The buffer has the same capacity as the byte array length, and its write offset is placed at the
+     * end, and its read offset is at the beginning, such that the entire buffer contents are readable.
+     * <p>
+     * The buffers produced by the supplier will each have their own independent life-cycle, and closing them will
+     * make them {@linkplain Buffer#isAccessible() inaccessible}, just like normally allocated buffers.
+     * <p>
+     * The buffers produced are "constants", in the sense that they are {@linkplain Buffer#readOnly() read-only}.
+     * <p>
+     * It can generally be expected, but is not guaranteed, that the returned supplier is more resource efficient than
+     * allocating and copying memory with other available APIs. In such optimised implementations, the underlying memory
+     * baking the buffers will be shared among all the buffers produced by the supplier.
+     * <p>
+     * The primary use case for this API, is when you need to repeatedly produce buffers with the same contents, and
+     * you perhaps wish to keep a {@code static final} field with these contents. The supplier-based API enforces
+     * that each usage get their own distinct buffer instance. Each of these instances cannot interfere with each other,
+     * so bugs like closing, or modifying the contents, of a shared buffer cannot occur.
+     *
+     * @param bytes The byte contents of the buffers produced by the returned supplier.
+     * @return A supplier of read-only buffers with the given contents.
+     */
+    default Supplier<Buffer> constBufferSupplier(byte[] bytes) {
+        byte[] safeCopy = bytes.clone(); // Prevent modifying the bytes after creating the supplier.
+        return () -> allocate(bytes.length).writeBytes(safeCopy).makeReadOnly();
```

Comment:
I don't think this should make a byte array copy each time. Wouldn't it be better to make the copy once, and then return a split()  ?

Reply:
The `ManagedBufferAllocator` does something like that. We can remove the default implementation and force implementors to provide an efficient implementation.